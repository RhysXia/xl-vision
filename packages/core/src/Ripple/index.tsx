import React from 'react'
import PropTypes from 'prop-types'
import clsx from 'clsx'
import TransitionGroup, { TransitionGroupClasses } from '../TransitionGroup'
import useEventCallback from '../commons/hooks/useEventCallback'
import createUseStyles from '../styles/createUseStyles'

export interface RippleProps extends React.HTMLAttributes<HTMLDivElement> {
  transitionClasses?: TransitionGroupClasses
  leaveAfterEnter?: boolean
  className?: string
  classes?: Partial<{
    root: string
    inner: string
  }>
}

export interface RippleRef {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  start: (e?: any) => void
  stop: () => void
}

const DELAY_RIPPLE = 80

const Ripple = React.forwardRef<RippleRef, RippleProps>((props, ref) => {
  const { transitionClasses, leaveAfterEnter, className, classes, ...others } = props

  const builtinStyles = useStyles(classes)

  const [ripples, setRipples] = React.useState<Array<React.ReactElement>>([])
  const finishedCountRef = React.useRef(0)
  const waitFinishedCountRef = React.useRef(0)
  const keyRef = React.useRef(0)
  const ignoreMouseDonwRef = React.useRef(false)
  const containerRef = React.useRef<HTMLDivElement>(null)

  const startTimerRef = React.useRef<NodeJS.Timeout>()

  React.useEffect(() => {
    return () => {
      startTimerRef.current && clearTimeout(startTimerRef.current)
    }
  }, [])

  const startCommit = React.useCallback(
    ({ x, y, size }: { x: number; y: number; size: number }) => {
      const key = keyRef.current
      const style: React.CSSProperties = {
        width: size,
        height: size,
        top: -size / 2 + y,
        left: -size / 2 + x
      }
      const ripple = <div className={builtinStyles.inner} key={key} style={style} />
      setRipples((prev) => [...prev, ripple])
      keyRef.current++
    },
    [builtinStyles.inner]
  )

  const start = React.useCallback(
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    (e: React.SyntheticEvent | object = {}) => {
      if (startTimerRef.current) {
        clearTimeout(startTimerRef.current)
        startTimerRef.current = undefined
      }

      if ((e as React.MouseEvent).type === 'mousedown' && ignoreMouseDonwRef.current) {
        ignoreMouseDonwRef.current = false
        return
      }

      if ((e as React.TouchEvent).type === 'touchstart') {
        ignoreMouseDonwRef.current = true
      }

      const el = (e as React.UIEvent).currentTarget
        ? ((e as React.UIEvent).currentTarget as HTMLElement)
        : containerRef.current
      const rect = el
        ? el.getBoundingClientRect()
        : {
            left: 0,
            top: 0,
            width: 0,
            height: 0
          }

      const { clientX, clientY } = (e as React.TouchEvent<HTMLDivElement>).touches
        ? (e as TouchEvent).touches[0]
        : (e as MouseEvent)

      const clientWidth = el ? el.clientWidth : 0
      const clientHeight = el ? el.clientHeight : 0
      const x = Math.round(typeof clientX === 'undefined' ? clientWidth / 2 : clientX - rect.left)
      const y = Math.round(typeof clientY === 'undefined' ? clientHeight / 2 : clientY - rect.top)
      const sizeX = Math.max(Math.abs(clientWidth - x), x) * 2 + 2
      const sizeY = Math.max(Math.abs(clientHeight - y), y) * 2 + 2
      const size = Math.round(Math.sqrt(sizeX ** 2 + sizeY ** 2))

      if ((e as React.TouchEvent).touches) {
        startTimerRef.current = setTimeout(() => {
          startCommit({ x, y, size })
        }, DELAY_RIPPLE)
      } else {
        startCommit({ x, y, size })
      }
    },
    [startCommit]
  )

  const stop = useEventCallback(() => {
    if (leaveAfterEnter) {
      if (finishedCountRef.current > 0) {
        setRipples((prev) => prev.slice(1))
        finishedCountRef.current--
        return
      }

      // 当多次触发stop，需要判断是否触发次数过多
      if (ripples.length > 0) {
        waitFinishedCountRef.current++
      }
    } else {
      setRipples((prev) => {
        if (prev.length > 0) {
          return prev.slice(1)
        }
        return prev
      })
    }
  })

  const afterEnter = useEventCallback(() => {
    if (leaveAfterEnter) {
      if (waitFinishedCountRef.current > 0) {
        setRipples((prev) => prev.slice(1))
        waitFinishedCountRef.current--
        return
      }
      finishedCountRef.current++
    }
  })

  React.useImperativeHandle(ref, () => ({
    start,
    stop
  }))

  return (
    <span {...others} className={clsx(builtinStyles.root, className)} ref={containerRef}>
      <TransitionGroup transitionClasses={transitionClasses} afterEnter={afterEnter}>
        {ripples}
      </TransitionGroup>
    </span>
  )
})

Ripple.displayName = 'Ripple'

Ripple.propTypes = {
  leaveAfterEnter: PropTypes.bool,
  className: PropTypes.string,
  classes: PropTypes.object,
  transitionClasses: PropTypes.oneOfType([PropTypes.string, PropTypes.object])
}

export default Ripple

const useStyles = createUseStyles({
  root: {
    position: 'absolute',
    top: 0,
    right: 0,
    bottom: 0,
    left: 0,
    zIndex: 0,
    overflow: 'hidden',
    borderRadius: 'inherit',
    pointerEvents: 'none'
  },
  inner: {
    position: 'absolute',
    backgroundColor: 'currentColor',
    borderRadius: '50%'
  }
})
