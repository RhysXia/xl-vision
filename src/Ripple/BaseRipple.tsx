import React from 'react'
import CssTransition, { CssTransitionClassNames } from '../CssTransition'

export interface BaseRippleProps {
  classNames: CssTransitionClassNames
  in: boolean
  size: number
  x: number
  y: number
  onExit: () => void
}

const Ripple: React.FunctionComponent<BaseRippleProps> = (props) => {
  const { in: inProp = false, classNames, size, x, y, onExit } = props

  const styles: React.CSSProperties = {
    width: size,
    height: size,
    top: -size / 2 + y,
    left: -size / 2 + x
  }

  return (
    <CssTransition in={inProp} classNames={classNames} afterEnter={onExit}>
      <span style={styles}></span>
    </CssTransition>
  )
}

export default Ripple
