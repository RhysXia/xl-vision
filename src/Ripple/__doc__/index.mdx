# Ripple

非常方便地实现点击时涟漪动画效果

:::demo 基本用法

将 Ripple 放在想要使用涟漪效果的容器中，配合一定的样式就可以很方便地实现点击时涟漪动画效果

```jsx
import React from 'react'
import { Ripple } from 'xl-vision'

export default () => {
  return (
    <div className='box'>
      <Ripple transitionClasses='ripple' />
    </div>
  )
}
```

```scss
.box {
  position: relative;
  height: 100px;
  background-color: #eee;
  color: #80bdff;

  &::deep {
    .ripple-enter-active,
    .ripple-leave-active {
      transition: all 0.4s ease;
    }
    .ripple-enter,
    .ripple-leave-to {
      transform: scale(0);
      opacity: 0.1;
    }
    .ripple-enter-to,
    .ripple-leave {
      transform: scale(1);
      opacity: 0.3;
    }
  }
}
```

:::

## Ripple API

| 参数 | 说明 | 类型 | 可选值 | 默认值 | 是否必填 |
| --- | --- | --- | --- | --- | --- |
| clsPrefix | 自定义类名前缀 | string | - | - | false |
| transitionClasses | 自定义动画过渡class | string &#x007c; object | - | - | false |
