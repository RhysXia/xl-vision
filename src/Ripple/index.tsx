import React, { TouchEvent, MouseEvent } from 'react'
import TransitionGroup, { TransitionGroupClasses } from '../TransitionGroup'
import ConfigContext from '../ConfigProvider/ConfigContext'
import useConstantCallback from '../commons/hooks/useConstantCallback'
import PropTypes from 'prop-types'
import classnames from 'classnames'

export interface RippleProps extends React.HTMLAttributes<HTMLDivElement> {
  transitionClasses?: TransitionGroupClasses
  clsPrefix?: string
  leaveAfterEnter?: boolean
  disableEvents?: boolean
}

export interface RippleRef {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  start: (e: any) => void
  stop: () => void
}

const Ripple = React.forwardRef<RippleRef, RippleProps>((props, ref) => {
  const { clsPrefix: rootClsPrefix } = React.useContext(ConfigContext)
  const {
    disableEvents,
    clsPrefix = `${rootClsPrefix}-ripple`,
    transitionClasses,
    leaveAfterEnter,
    className,
    ...others
  } = props

  const [ripples, setRipples] = React.useState<Array<React.ReactElement>>([])
  const finishedCountRef = React.useRef(0)
  const waitFinishedCountRef = React.useRef(0)
  const keyRef = React.useRef(0)
  const ignoreMouseDonwRef = React.useRef(false)
  const containerRef = React.useRef<HTMLDivElement>(null)

  const startCommit = React.useCallback(
    ({ x, y, size }: { x: number; y: number; size: number }) => {
      const key = keyRef.current
      const styles: React.CSSProperties = {
        width: size,
        height: size,
        top: -size / 2 + y,
        left: -size / 2 + x
      }
      const ripple = <div className={`${clsPrefix}__inner`} key={key} style={styles}></div>
      setRipples((prev) => [...prev, ripple])
      keyRef.current++
    },
    [clsPrefix]
  )

  const start = React.useCallback(
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    (e: any = {}) => {
      if (e.type === 'mousedown' && ignoreMouseDonwRef.current) {
        ignoreMouseDonwRef.current = false
        return
      }

      if (e.type === 'touchstart') {
        ignoreMouseDonwRef.current = true
      }

      const el = e.currentTarget ? (e.currentTarget as HTMLElement) : containerRef.current
      const rect = el
        ? el.getBoundingClientRect()
        : {
            left: 0,
            top: 0,
            width: 0,
            height: 0
          }

      const { clientX, clientY } = (e as React.TouchEvent<HTMLDivElement>).touches
        ? (e as TouchEvent).touches[0]
        : (e as MouseEvent)

      const clientWidth = el ? el.clientWidth : 0
      const clientHeight = el ? el.clientHeight : 0
      const x = Math.round(typeof clientX === 'undefined' ? clientWidth / 2 : clientX - rect.left)
      const y = Math.round(typeof clientY === 'undefined' ? clientHeight / 2 : clientY - rect.top)
      const sizeX = Math.max(Math.abs(clientWidth - x), x) * 2 + 2
      const sizeY = Math.max(Math.abs(clientHeight - y), y) * 2 + 2
      const size = Math.round(Math.sqrt(sizeX ** 2 + sizeY ** 2))
      startCommit({ x, y, size })
    },
    [startCommit]
  )

  const stop = useConstantCallback(() => {
    if (leaveAfterEnter) {
      if (finishedCountRef.current > 0) {
        setRipples((prev) => prev.slice(1))
        finishedCountRef.current--
        return
      }

      // 当多次触发stop，需要判断是否触发次数过多
      if (ripples.length > 0) {
        waitFinishedCountRef.current++
      }
    } else {
      setRipples((prev) => {
        if (prev.length > 0) {
          return prev.slice(1)
        }
        return prev
      })
    }
  })

  const afterEnter = useConstantCallback(() => {
    if (leaveAfterEnter) {
      if (waitFinishedCountRef.current > 0) {
        setRipples((prev) => prev.slice(1))
        waitFinishedCountRef.current--
        return
      }
      finishedCountRef.current++
    }
  })

  React.useImperativeHandle(ref, () => ({
    start,
    stop
  }))

  const events = React.useMemo(() => {
    if (disableEvents) {
      return
    }
    return {
      onTouchStart: start,
      onTouchEnd: stop,
      onTouchMove: stop,
      onMouseDown: start,
      onMouseUp: stop,
      onMouseLeave: stop,
      onDragLeave: stop
    }
  }, [disableEvents, start, stop])

  const classes = classnames(clsPrefix, className)

  return (
    <div {...others} {...events} className={classes} ref={containerRef}>
      <TransitionGroup transitionClasses={transitionClasses} afterEnter={afterEnter}>
        {ripples}
      </TransitionGroup>
    </div>
  )
})

Ripple.displayName = 'Ripple'

Ripple.propTypes = {
  disableEvents: PropTypes.bool,
  clsPrefix: PropTypes.string,
  leaveAfterEnter: PropTypes.bool,
  className: PropTypes.string,
  transitionClasses: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.shape({
      enter: PropTypes.string,
      enterActive: PropTypes.string,
      enterTo: PropTypes.string,
      leave: PropTypes.string,
      leaveActive: PropTypes.string,
      leaveTo: PropTypes.string,
      move: PropTypes.string
    })
  ])
}

export default Ripple
