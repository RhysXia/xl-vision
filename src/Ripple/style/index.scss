@use './var' as *;

$ripple: #{$cls-prefix}-ripple;

.#{$ripple} {
  position: absolute;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  overflow: hidden;
  border-radius: inherit;
}
