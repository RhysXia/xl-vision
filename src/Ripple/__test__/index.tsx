import { mount } from 'enzyme'
import Ripple from '..'
import React from 'react'

describe('Ripple', () => {
  it('基本测试', () => {
    const wrapper = mount(
      <div>
        click me
        <Ripple transitionClasses='ripple' />
      </div>
    )

    expect(wrapper.render()).toMatchSnapshot()

    wrapper.find('.xl-ripple').simulate('mousedown')
    wrapper.update()

    expect(wrapper.render()).toMatchSnapshot()

    wrapper.find('.xl-ripple').simulate('mouseup')
    wrapper.update()
    expect(wrapper.render()).toMatchSnapshot()
  })

  it('测试disableEvents', () => {
    const wrapper = mount(
      <div>
        click me
        <Ripple disableEvents={true} transitionClasses='ripple' />
      </div>
    )

    expect(wrapper.render()).toMatchSnapshot()

    wrapper.find('.xl-ripple').simulate('mousedown')
    wrapper.update()

    expect(wrapper.render()).toMatchSnapshot()

    wrapper.find('.xl-ripple').simulate('mouseup')
    wrapper.update()
    expect(wrapper.render()).toMatchSnapshot()
  })
})
