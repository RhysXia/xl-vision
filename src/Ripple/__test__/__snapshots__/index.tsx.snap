// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Ripple 基本测试 1`] = `
<div>
  click me
  <span
    class="xl-ripple"
  />
</div>
`;

exports[`Ripple 基本测试 2`] = `
<div>
  click me
  <span
    class="xl-ripple"
  >
    <div
      class="xl-ripple__inner ripple-enter ripple-enter-active"
      style="width: 3px; height: 3px; top: -1.5px; left: -1.5px;"
    />
  </span>
</div>
`;

exports[`Ripple 基本测试 3`] = `
<div>
  click me
  <span
    class="xl-ripple"
  >
    <div
      class="xl-ripple__inner ripple-leave ripple-leave-active"
      style="width: 3px; height: 3px; top: -1.5px; left: -1.5px;"
    />
  </span>
</div>
`;

exports[`Ripple 测试disableEvents 1`] = `
<div>
  click me
  <span
    class="xl-ripple"
  />
</div>
`;

exports[`Ripple 测试disableEvents 2`] = `
<div>
  click me
  <span
    class="xl-ripple"
  />
</div>
`;

exports[`Ripple 测试disableEvents 3`] = `
<div>
  click me
  <span
    class="xl-ripple"
  />
</div>
`;
